src/libraries/Common/tests/System/Net/Sockets/TestPortPool.cs

```diff
@@ -0,0 +1,134 @@
+    }
+
+    internal class TestPortPoolExhaustedException : Exception
+    {
+        public TestPortPoolExhaustedException()
+            : base($"TestPortPool failed to find an available port after {TestPortPool.ThrowExhaustedAfter} attempts")
+        {
+        }
+    }
+
+    /// <summary>
+    /// Distributes unique ports from the range defined by <see cref="Configuration.Sockets.TestPoolPortRange"/>
+    /// Useful in socket testing scenarios, where port collisions across protocols are not acceptable.
+    /// This kind of uniqueness is not guaranteed when binding to OS ephemeral ports, and might lead to issues on Unix.
+    /// For more information see:
+    /// https://github.com/dotnet/runtime/issues/19162#issuecomment-523195762
+    /// </summary>
+    internal static class TestPortPool
+    {
+        internal const int ThrowExhaustedAfter = 10;
+
+        private static readonly int MinPort =
+            System.Net.Test.Common.Configuration.Sockets.TestPoolPortRange.Min;
+        private static readonly int MaxPort =
+            System.Net.Test.Common.Configuration.Sockets.TestPoolPortRange.Max;
+        private static readonly int PortRangeLength = MaxPort - MinPort;
+
+        private static readonly ConcurrentDictionary<int, int> s_usedPorts = GetAllPortsUsedBySystem();
+        private static int s_counter = int.MinValue;
+
+        public static PortLease RentPort()
+        {
+            for (int i = 0; i < ThrowExhaustedAfter; i++)
+            {
+                // Although race may occur theoretically because the following code block is not atomic,
+                // it requires the s_counter to move at least PortRangeLength steps between Increment and TryAdd,
+                // which is very unlikely considering the actual port range and the low number of tests utilizing TestPortPool
+
+                long portLong = (long)Interlocked.Increment(ref s_counter) - int.MinValue;
+                portLong = (portLong % PortRangeLength) + MinPort;
+                int port = (int)portLong;
+
+                if (s_usedPorts.TryAdd(port, 0))
+                {
+                    return new PortLease(port);
+                }
+            }
+
+            throw new TestPortPoolExhaustedException();
+        }
+
+        public static void Return(PortLease portLease)
+        {
+            s_usedPorts.TryRemove(portLease.Port, out _);
+        }
+
+        public static PortLease RentPortAndBindSocket(Socket socket, IPAddress address)
+        {
+            PortLease lease = RentPort();
+            try
+            {
+                socket.Bind(new IPEndPoint(address, lease.Port));
+                return lease;
+            }
+            catch (SocketException)
+            {
+                lease.Dispose();
+                throw;
+            }
+        }
+
+        // Exclude ports which are unavailable at initialization time
+        private static ConcurrentDictionary<int, int> GetAllPortsUsedBySystem()
```

Comment:
What if any of the ports in the allowed range are bound by other tests and/or processes after the static initializer runs?

Kestrel tests used to check if certain ports were available during xunit test discovery, but would find that the port would become unavailable by the time the test ran. Now, for tests that want to verify binding to a specific port, we test if the port is available at the start of each test, and skip the test otherwise. This improved reliability, but even that logic is still sometimes flaky, and that's despite running those tests serially in their own test group so they shouldn't be running in parallel with other tests.

Can we have RentPort() redo all the IsPortUsed() checks at the last possible moment in order to double check nothing else has taken the port? I think RentPortAndBindSocket should also retry on bind failures instead of throw.

Reply:
The configured port range for `TestPortPool` does not overlap with the OS ephemeral ports. I think if we make sure that this range s dedicated to the tests using `TestPortPool`, and no other tests try to arbitrarily bind to it's ports, we don't need to be concerned about this in System.Net.Sockets.Tests. We may add a README.MD, to make sure no one will violate this in the future. Or am I missing something?

I wanted to avoid the creation + destroyal of temporary socket handles, because I was afraid of possible esoteric OS side effects I'm not aware of. We need to create & destroy 4 sockets per rental, according to the "unique-across-protocols" contract guaranteed by `TestPortPool`. This might be a false concern though.